tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,300 @@
+#![cfg_attr(not(feature = "sync"), allow(dead_code, unreachable_pub))]
+
+use super::Semaphore;
+use crate::loom::cell::UnsafeCell;
+use std::cell::Cell;
+use std::fmt;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A thread-safe cell which can be written to only once.
+///
+/// Provides the functionality to set the value by using an async
+/// function via [`OnceCell::get_or_init`]
+///
+/// [`OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
+///
+/// # Examples
+/// ```
+/// use tokio::sync::OnceCell;
+/// use std::pin::Pin;
+/// use std::future::Future;
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// fn call_async_fn() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+///     Box::pin(some_computation())
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     static ONCE: OnceCell<u32> = OnceCell::new();
+///     let result1 = tokio::spawn(async {
+///         ONCE.get_or_init(call_async_fn).await
+///     }).await.unwrap();
+///     assert_eq!(*result1, 2);
+/// }
+pub struct OnceCell<T> {
+    value_set: AtomicBool,
+    value: UnsafeCell<MaybeUninit<T>>,
+    sema: Semaphore,
+}
+
+impl<T> Default for OnceCell<T> {
+    fn default() -> OnceCell<T> {
+        OnceCell::new()
+    }
+}
+
+impl<T> fmt::Debug for OnceCell<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("OnceCell")
+            .field("value", &self.value)
+            .finish()
+    }
+}
+
+impl<T: Clone> Clone for OnceCell<T> {
+    fn clone(&self) -> OnceCell<T> {
+        let new_cell = OnceCell::new();
+        if let Ok(value) = self.get() {
+            match new_cell.set(value.clone()) {
+                Ok(()) => (),
+                Err(_) => unreachable!(),
+            }
+        }
+        new_cell
+    }
+}
+
+impl<T: PartialEq> PartialEq for OnceCell<T> {
+    fn eq(&self, other: &OnceCell<T>) -> bool {
+        self.get() == other.get()
+    }
+}
+
+impl<T: Eq> Eq for OnceCell<T> {}
+
+impl<T> OnceCell<T> {
+    /// Creates a new uninitialized OnceCell instance.
+    pub const fn new() -> Self {
+        OnceCell {
+            value_set: AtomicBool::new(false),
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            sema: Semaphore::const_new(1),
+        }
+    }
+
+    fn initialized(&self) -> bool {
+        self.value_set.load(Ordering::Acquire)
+    }
+
+    fn set_to_initialized(&self) {
+        self.value_set.store(true, Ordering::Release);
+    }
+
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+    }
```

Comment:
This method could totally include the `value_set = true` and closing the semaphore.